src/libPMacc/include/particles/ParticlesBase.kernel

```diff
@@ -49,181 +52,301 @@ getPreviousFrameAndRemoveLastFrame( const typename T_ParticleBox::FramePtr& fram
-        const uint32_t numExchanges = traits::NumberOfExchanges<dim>::value;
+        constexpr uint32_t numExchanges = traits::NumberOfExchanges< dim >::value;
+        constexpr uint32_t numWorker = T_worker;
 
         /* define memory for two times Exchanges
          * index range [0,numExchanges-1] are being referred to as `low frames`
          * index range [numExchanges,2*numExchanges-1] are being referred to as `high frames`
          */
-        PMACC_SMEM( destFrames, memory::Array< FramePtr, numExchanges * 2 > );
+        PMACC_SMEM(
+            destFrames,
+            memory::Array<
+                FramePtr,
+                numExchanges * 2
+            >
+        );
         //count particles per frame
-        PMACC_SMEM( destFramesCounter, memory::Array< int, numExchanges > );
-
-        /* lastFrameSize is only valid for threads with `linearThreadIdx` < numExchanges */
-        uint32_t lastFrameSize = 0;
-
-        PMACC_SMEM( frame, FramePtr );
-        PMACC_SMEM( mustShift, bool );
-
-        DataSpace<dim> superCellIdx = mapper.getSuperCellIndex( DataSpace<dim> (blockIdx) );
-        const uint32_t linearThreadIdx = threadIdx.x;
-
-        if ( linearThreadIdx == 0 )
-        {
-            mustShift = pb.getSuperCell( superCellIdx ).mustShift( );
-            if ( mustShift )
+        PMACC_SMEM(
+            destFramesCounter,
+            memory::Array<
+                int,
+                numExchanges
+            >
+        );
+
+        PMACC_SMEM(
+            frame,
+            FramePtr
+        );
+        PMACC_SMEM(
+            mustShift,
+            bool
+        );
+
+        DataSpace< dim > superCellIdx = mapper.getSuperCellIndex( DataSpace< dim >( blockIdx ) );
+        uint32_t const workerIdx = threadIdx.x;
+
+        using namespace mappings::threads;
+
+        ForEachIdx<
+            IdxConfig<
+                1,
+                numWorker
+            >
+        >{ workerIdx }(
+            [&](
+                uint32_t const,
+                uint32_t const
+            )
             {
-                pb.getSuperCell( superCellIdx ).setMustShift( false );
-                frame = pb.getFirstFrame( superCellIdx );
+                mustShift = pb.getSuperCell( superCellIdx ).mustShift( );
+                if ( mustShift )
+                {
+                    pb.getSuperCell( superCellIdx ).setMustShift( false );
+                    frame = pb.getFirstFrame( superCellIdx );
+                }
             }
-        }
+        );
 
         __syncthreads( );
         if ( !mustShift || !frame.isValid( ) ) return;
 
-        const DataSpace<dim> relative = superCellIdx + Mask::getRelativeDirections<dim> (linearThreadIdx + 1);
+        using ExchangeDomCfg = IdxConfig<
+            numExchanges,
+            numWorker
+        >;
+
+        memory::CtxArray<
+            uint32_t,
+            ExchangeDomCfg
+        > lastFrameSizeCtx( 0 );
+
```

Comment:
Why is `lastFrameSize`(`Ctx`) not a index local variable anymore? Do you need this this after the `forEach`? Do you assume any optimizations benefits here, I can see? If yes, it would be helpful to comment them. :wink: